airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionHandler.kt

```diff
@@ -0,0 +1,175 @@
+        e: Throwable,
+        cmd: Command,
+        outputRecordCollector: Consumer<AirbyteMessage>
+    ) {
+        ApmTraceUtils.addExceptionToTrace(e)
+        val rootException: Throwable? = getRootException(e)
+        val externalMessage: String? = getExternalMessage(rootException)
+        // error messages generated during check() needs special handling
+        if (cmd == Command.CHECK) {
+            outputRecordCollector.accept(
+                AirbyteMessage()
+                    .withType(AirbyteMessage.Type.CONNECTION_STATUS)
+                    .withConnectionStatus(
+                        AirbyteConnectionStatus()
+                            .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                            .withMessage(externalMessage),
+                    ),
+            )
+        } else {
+            if (checkErrorType(rootException, FailureType.CONFIG)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(e, externalMessage)
+                exitProcess(1)
+            } else if (checkErrorType(rootException, FailureType.TRANSIENT)) {
+                AirbyteTraceMessageUtility.emitTransientErrorTrace(e, externalMessage)
+                exitProcess(1)
+            }
+            throw e
+        }
+    }
+
+    /**
+     * Initializes the error dictionary for the connector. This method shall include all the errors
+     * that are shared by all connectors. For instance, Debezium errors
+     */
+    open fun initializeErrorDictionary() {}
+
+    /**
+     * Translates an internal exception message to an external user-friendly message. This is the
+     * main entrance of the error translation process.
+     */
+    fun getExternalMessage(e: Throwable?): String? {
+        // some common translations that every connector would share can be done here
+        if (e is ConfigErrorException) {
+            return e.displayMessage
+        } else if (e is TransientErrorException) {
+            return e.message
+        } else if (e is ConnectionErrorException) {
+            return ErrorMessage.getErrorMessage(e.stateCode, e.errorCode, e.exceptionMessage, e)
+        } else {
+            val msg = translateConnectorSpecificErrorMessage(e)
+            if (msg != null) return msg
+        }
+        // if no specific translation is found, return a generic message
+        return String.format(
+            COMMON_EXCEPTION_MESSAGE_TEMPLATE,
+            if (e!!.message != null) e.message else "",
+        )
+    }
+
+    /**
+     * Translates a connector specific error message to an external user-friendly message. This
+     * method should be implemented by individual connectors that wish to translate connector
+     * specific error messages.
+     */
+    open fun translateConnectorSpecificErrorMessage(e: Throwable?): String? {
+        for (error in connectorErrorDictionary) {
+            if (e!!.message?.lowercase()?.matches(error.regexMatchingPattern.toRegex())!!)
```

Comment:
is this !! safe? `rootException` can be null

edit: or can it? please tighten the types in your method arguments and deal with the `null`s early on

Reply:
agreed, fixed.